tests/PHPUnit/Integration/Session/SessionAuthTest.php

```diff
@@ -0,0 +1,165 @@
+        $this->initializeRequest(self::TEST_IP, 'some-other-user-agent', Fixture::ADMIN_USER_LOGIN);
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::FAILURE, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsSuccess_IfRequestIpAndUserAgentMatchSession_AndCookieUsernameMatchesSessionUsername()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::SUCCESS, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsSuperUserSuccess_IfRequestIpAndUserAgentMatchSession_AndCookieUsernameMatchesSessionUsername()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::SUCCESS_SUPERUSER_AUTH_CODE, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsFailure_IfNoSessionExists()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+
+        $this->destroySession();
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::FAILURE, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsFailure_IfAuthenticatedSession_AndPasswordChangedAfterSessionCreated()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+
+        sleep(1);
+
+        UsersManagerAPI::getInstance()->updateUser(self::TEST_OTHER_USER, 'testpass2');
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::FAILURE, $result->getCode());
+
+        $this->assertEmpty($_SESSION);
+    }
+
+    private function initializeRequest($ip, $userAgent, $userName, $passwordModifiedTime = false)
+    {
+        if (empty($passwordModifiedTime)) {
+            $model = new Model();
+            $user = $model->getUser($userName);
+            $passwordModifiedTime = $user['ts_password_modified'];
+
+            if (empty($passwordModifiedTime)) { // sanity check
+                throw new \Exception('ts_password_modified is empty');
+            }
+        }
+
+        $sessionCookieFactory = StaticContainer::get(SessionAuthCookieFactory::class);
+        $sessionFingerprint = new SessionFingerprint();
+
+        $cookie = $sessionCookieFactory->getCookie(false);
+        $cookie->set('id', $sessionFingerprint->getHash($passwordModifiedTime));
```

Comment:
Any reason for not letting PHP set the session cookie? and generate the ID?

Reply:
The existing code uses Cookie & overrides some parts of Zend_Session. Didn't want to just get rid of all of that in case there were reasons for its existence, but I can if that's desired.

Reply:
Reread this question and there's some misconception about the 'id' in the cookie. Originally that was the hash of the hash of the token auth (or something like that). I'm not sure if 'id' is the right word for it.

Anyway, I removed the session secret, so now the piwik_auth cookie is empty. It is, however, required (at the moment anyway). In FrontController we check if this cookie is sent and if so, we use SessionAuth instead of the normal Auth implementation.

We can't use the PIWIK_SESSID cookie (which is the PHP session ID) since in the UI there is always a session, even when you're logged out. I believe this is for form nonces (like the one used in the login form), which are stored in sessions.

I thought about checking if the session is authenticated (ie, the 'user.name' variable exists in the session), but that would mean always doing session_start(), and to my understanding, we don't want to do that for, eg, every API request, just for requests from a browser.